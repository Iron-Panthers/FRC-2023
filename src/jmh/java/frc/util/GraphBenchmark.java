package frc.util;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;

@State(Scope.Benchmark)
public class GraphBenchmark {
  private static String[] nodes = {
    "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S",
    "T", "U", "V", "W", "X", "Y", "Z"
  };

  @Benchmark
  public Graph<String> constructGraph() {
    Graph<String> graph = new Graph<>();
    for (String node : nodes) {
      graph.addNode(node);
    }
    for (int i = 0; i < nodes.length; i++) {
      for (int j = i + 1; j < nodes.length; j++) {
        graph.addEdge(nodes[i], nodes[j], i * j);
      }
    }
    graph.lock();
    return graph;
  }
}
