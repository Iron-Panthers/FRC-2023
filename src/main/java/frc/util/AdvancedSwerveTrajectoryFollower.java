// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.util;

import com.pathplanner.lib.PathPlannerTrajectory.PathPlannerState;
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.ProfiledPIDController;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.trajectory.Trajectory;
import frc.robot.autonomous.TrajectoryFollower;

/**
 * Implements a simple swerve trajectory follower, with a fixed robot heading.
 *
 * <p>See TrajectoryFollower class for more detailed documentation.
 */
public class AdvancedSwerveTrajectoryFollower extends TrajectoryFollower<ChassisSpeeds> {
  private final PIDController xController;
  private final PIDController yController;
  private final ProfiledPIDController angleController;

  private Trajectory.State lastState = null;
  private boolean finished = false;
  /* Variable to track if calculateDriveSignal has run once yet */
  private boolean firstRun = false;

  public AdvancedSwerveTrajectoryFollower(
      PIDController xController, PIDController yController, ProfiledPIDController angleController) {
    this.xController = xController;
    // var tab = Shuffleboard.getTab("tuneFollower");
    // tab.add(xController);
    // tab.add(yController);
    // tab.addDouble("xError", xController::getPositionError);
    // tab.addDouble("yError", yController::getPositionError);
    this.yController = yController;
    this.angleController = angleController;
  }

  private ChassisSpeeds finishTrajectory() {
    finished = true;
    return new ChassisSpeeds();
  }

  @Override
  protected ChassisSpeeds calculateDriveSignal(
      Pose2d currentPose, Trajectory trajectory, double time, double dt) {
    if (time > trajectory.getTotalTimeSeconds()) {
      // If the robot is past the end of the trajectory, stop
      return finishTrajectory();
    }

    // there is still time left!
    lastState = trajectory.sample(time);
    if (firstRun) {
      angleController.reset(currentPose.getRotation().getRadians());
      firstRun = false;
    }
    Pose2d poseRef = lastState.poseMeters;
    double linearVelocityRefMeters = lastState.velocityMetersPerSecond;
    double xFF = linearVelocityRefMeters * poseRef.getRotation().getCos();
    double yFF = linearVelocityRefMeters * poseRef.getRotation().getSin();

    double currentDegrees = (360 - currentPose.getRotation().getDegrees()) % 360;
    double targetDegrees =
        // only cast if safe
        lastState instanceof PathPlannerState
            ? ((360 - ((PathPlannerState) lastState).holonomicRotation.getDegrees()) % 360)
            : ((360 - poseRef.getRotation().getDegrees()) % 360);

    // scope current and target angles
    double angularDifferenceDeg = Util.relativeAngularDifference(currentDegrees, targetDegrees);

    // SmartDashboard.putNumber("current pose rotation", currentDegrees);
    // SmartDashboard.putNumber("pose ref rotation", targetDegrees);
    // SmartDashboard.putNumber("sstf/angular_difference (deg)", angularDifferenceDeg);

    // use pid controller using scoped angles to get shortest-distance correction
    double angleFF = angleController.calculate(angularDifferenceDeg, 0);

    // SmartDashboard.putNumber("angleff", angleFF);

    double xControllerEffort = xController.calculate(currentPose.getX(), poseRef.getX());
    double yControllerEffort = yController.calculate(currentPose.getY(), poseRef.getY());

    return ChassisSpeeds.fromFieldRelativeSpeeds(
        xFF + xControllerEffort, yFF + yControllerEffort, angleFF, currentPose.getRotation());
  }

  public Trajectory.State getLastState() {
    return lastState;
  }

  @Override
  protected boolean isFinished() {
    return finished;
  }

  @Override
  protected void reset() {
    xController.reset();
    yController.reset();
    angleController.reset(0);
    finished = false;
    firstRun = false;
  }
}
