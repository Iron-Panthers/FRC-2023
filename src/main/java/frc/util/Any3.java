package frc.util;

import java.util.function.Function;
import java.util.function.Supplier;

public class Any3<A, B, C> {

  enum Type {
    A,
    B,
    C
  }

  private final A a;
  private final B b;
  private final C c;

  private final Type type;

  private Any3(A a, B b, C c, Type type) {
    this.a = a;
    this.b = b;
    this.c = c;
    this.type = type;
  }

  public static <A, B, C> Any3<A, B, C> ofA(A a) {
    return new Any3<>(a, null, null, Type.A);
  }

  public static <A, B, C> Any3<A, B, C> ofB(B b) {
    return new Any3<>(null, b, null, Type.B);
  }

  public static <A, B, C> Any3<A, B, C> ofC(C c) {
    return new Any3<>(null, null, c, Type.C);
  }

  public <R> Supplier<R> map(Function<A, R> aFn, Function<B, R> bFn, Function<C, R> cFn) {
    return switch (type) {
      case A -> () -> aFn.apply(a);
      case B -> () -> bFn.apply(b);
      case C -> () -> cFn.apply(c);
    };
  }
}
