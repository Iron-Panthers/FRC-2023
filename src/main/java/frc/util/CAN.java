package frc.util;

import java.util.HashSet;
import java.util.Set;

public class CAN {
  private CAN() {}

  public record CANDevice(int id, String name) {}

  private static final Set<CANDevice> devices = new HashSet<>();
  private static final Set<Integer> ids = new HashSet<>();

  public static int at(int id, String name) {
    devices.add(new CANDevice(id, name));
    ids.add(id);
    return id;
  }

  public static Set<CANDevice> getDevices() {
    return devices;
  }

  public static Set<Integer> getIds() {
    return ids;
  }
}
