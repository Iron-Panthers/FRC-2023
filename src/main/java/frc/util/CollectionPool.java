package frc.util;

import java.util.ArrayDeque;
import java.util.function.Consumer;
import java.util.function.Supplier;

public class CollectionPool<T> {
  private final ArrayDeque<T> pool = new ArrayDeque<>();
  private final Supplier<T> init;
  private final Consumer<T> reset;

  public CollectionPool(Supplier<T> init, Consumer<T> reset) {
    this.init = init;
    this.reset = reset;
  }

  public T get() {
    if (pool.isEmpty()) {
      return init.get();
    }
    var collection = pool.pop();
    reset.accept(collection);
    return collection;
  }

  public void put(T t) {
    // we will wait to reset the collection until it is popped from the pool
    pool.push(t);
  }
}
