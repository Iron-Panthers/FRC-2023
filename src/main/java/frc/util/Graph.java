package frc.util;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * An asymmetric, weighted graph implementation, designed for high performance for path finding
 * applications.
 *
 * <p>There is not support for removing nodes or edges.
 */
public class Graph<T> {
  /** An edge in the graph. Stores the destination node and the weight of the edge. */
  public static class Edge<T> {
    public final T to;
    public final int weight;

    public Edge(T to, int weight) {
      this.to = to;
      this.weight = weight;
    }

    @Override
    public int hashCode() {
      return to.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
      if (obj == null) return false;
      if (!(obj instanceof Edge)) return false;
      Edge<?> other = (Edge<?>) obj;
      return to.equals(other.to);
    }
  }

  private Map<T, List<Edge<T>>> internalHashMap = new HashMap<>();

  private boolean locked = false;

  /** Creates a new graph. */
  public Graph() {}

  /** lock the graph such that no more mutations can occur. */
  public void lock() {
    internalHashMap.forEach(
        (key, value) -> internalHashMap.put(key, Collections.unmodifiableList(value)));
    internalHashMap = Collections.unmodifiableMap(internalHashMap);
    locked = true;
  }

  private void checkLocked() {
    if (locked) {
      throw new UnsupportedOperationException("Graph is locked");
    }
  }

  private void assertNodeExists(T node) {
    if (!internalHashMap.containsKey(node)) {
      throw new IllegalArgumentException("Node does not exist " + node.toString());
    }
  }

  private void assertNodeDoesNotExist(T node) {
    if (internalHashMap.containsKey(node)) {
      throw new IllegalArgumentException("Node already exists " + node.toString());
    }
  }

  private Edge<T> internalGetEdge(T from, T to) {
    assertNodeExists(from);
    assertNodeExists(to);
    for (Edge<T> edge : internalHashMap.get(from)) {
      if (edge.to.equals(to)) {
        return edge;
      }
    }
    return null;
  }

  private void assertEdgeExists(T from, T to) {
    if (internalGetEdge(from, to) == null) {
      throw new IllegalArgumentException(
          "Edge does not exist " + from.toString() + " -> " + to.toString());
    }
  }

  private void assertEdgeDoesNotExist(T from, T to) {
    if (internalGetEdge(from, to) != null) {
      throw new IllegalArgumentException(
          "Edge already exists " + from.toString() + " -> " + to.toString());
    }
  }

  /**
   * Returns the immutable version of a set, to prevent handing out the ability to modify the
   * internal data structure.
   *
   * @param <T> The type of the set.
   * @param set The set to make immutable.
   * @return The immutable version of the set.
   */
  private static <T> Set<T> view(Set<T> set) {
    return Collections.unmodifiableSet(set);
  }

  private static <T> List<T> view(List<T> list) {
    return Collections.unmodifiableList(list);
  }

  /**
   * Adds a node to the graph if it does not already exist.
   *
   * @param node the node to add
   */
  public void addNode(T node) {
    checkLocked();
    assertNodeDoesNotExist(node);
    internalHashMap.put(node, new ArrayList<>());
  }

  /**
   * Adds an edge to the graph.
   *
   * @param from the node the edge is from
   * @param to the node the edge is to
   * @param weight the weight of the edge
   */
  public void addEdge(T from, T to, int weight) {
    checkLocked();
    assertEdgeDoesNotExist(from, to);
    internalHashMap.get(from).add(new Edge<>(to, weight));
  }

  /**
   * Gets the weight of an edge.
   *
   * @param from the node the edge is from
   * @param to the node the edge is to
   * @return the weight of the edge
   */
  public int getEdgeWeight(T from, T to) {
    assertNodeExists(from);
    assertNodeExists(to);
    for (Edge<T> edge : internalHashMap.get(from)) {
      if (edge.to.equals(to)) {
        return edge.weight;
      }
    }
    throw new IllegalArgumentException(
        "Edge does not exist " + from.toString() + " -> " + to.toString());
  }

  /**
   * Gets the neighbors of a node.
   *
   * @param node the node to get the neighbors of
   * @return the neighbors of the node, as a list of edges
   */
  public List<Edge<T>> getNeighbors(T node) {
    assertNodeExists(node);
    return view(internalHashMap.get(node));
  }

  /**
   * Gets the nodes in the graph.
   *
   * @return the nodes in the graph
   */
  public Set<T> getNodes() {
    return view(internalHashMap.keySet());
  }

  public boolean hasNode(T node) {
    return internalHashMap.containsKey(node);
  }

  public boolean hasEdge(T from, T to) {
    if (!hasNode(from) || !hasNode(to)) return false;

    for (Edge<T> edge : internalHashMap.get(from)) {
      if (edge.to.equals(to)) {
        return true;
      }
    }

    return false;
  }
}
