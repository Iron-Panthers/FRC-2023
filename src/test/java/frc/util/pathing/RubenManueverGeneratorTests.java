package frc.util.pathing;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertIterableEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.pathplanner.lib.PathConstraints;
import com.pathplanner.lib.PathPlanner;
import com.pathplanner.lib.PathPlannerTrajectory;
import edu.wpi.first.math.Pair;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import frc.UtilParamTest;
import frc.UtilTest;
import frc.robot.Constants.Pathing;
import frc.util.Graph;
import frc.util.Graph.Edge;
import frc.util.pathing.DisplayFieldArray.FieldSquare;
import java.lang.reflect.Field;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

@ExtendWith({SnapshotExtension.class})
public class RubenManueverGeneratorTests {

  @UtilTest
  public void constructsWithoutError() {
    assertDoesNotThrow(RubenManueverGenerator::new);
  }

  @UtilTest
  public void unobstructedCordsAreDoublyLinked() {
    RubenManueverGenerator rubenManueverGenerator = new RubenManueverGenerator();

    // use java reflection to access adjacencyGraph
    Optional<Graph<GridCoord>> optAdjacencyGraph = Optional.empty();
    try {
      Field field = RubenManueverGenerator.class.getDeclaredField("adjacencyGraph");
      field.setAccessible(true);

      var obj = field.get(rubenManueverGenerator);

      optAdjacencyGraph =
          obj instanceof Graph<?> ? Optional.of((Graph<GridCoord>) obj) : Optional.empty();
    } catch (NoSuchFieldException | IllegalAccessException e) {
      e.printStackTrace();
    }

    if (optAdjacencyGraph.isEmpty()) {
      fail("optAdjacencyGraph is empty because reflection failed");
    }

    var adjacencyGraph = optAdjacencyGraph.get();

    var coords =
        List.of(
            // basic cardinals
            Pair.of(new Translation2d(5, 5), new Translation2d(5, 5 + Pathing.CELL_SIZE_METERS)),
            Pair.of(new Translation2d(5, 5), new Translation2d(5 + Pathing.CELL_SIZE_METERS, 5)),
            Pair.of(new Translation2d(5, 5), new Translation2d(5, 5 - Pathing.CELL_SIZE_METERS)),
            Pair.of(new Translation2d(5, 5), new Translation2d(5 - Pathing.CELL_SIZE_METERS, 5)),
            // straight line connectivity
            Pair.of(
                new Translation2d(5 + Pathing.CELL_SIZE_METERS, 5),
                new Translation2d(5 + Pathing.CELL_SIZE_METERS * 2, 5))
            // Pair.of(
            //     coord(5 + Pathing.CELL_SIZE_METERS * 2, 5),
            //     coord(5 + Pathing.CELL_SIZE_METERS * 3, 5))
            // brace holder for autoformatting
            );

    for (var coord : coords) {
      var firstGridCoord = new GridCoord(coord.getFirst());
      var secondGridCoord = new GridCoord(coord.getSecond());
      assertFalse(
          FieldObstructionMap.isInsideObstruction(coord.getFirst()),
          String.format("coord %s is inside an obstruction", coord.getFirst()));
      assertFalse(
          FieldObstructionMap.isInsideObstruction(coord.getSecond()),
          String.format("coord %s is inside an obstruction", coord.getSecond()));
      assertTrue(
          adjacencyGraph.hasNode(firstGridCoord),
          String.format("Node %s is missing", firstGridCoord));
      assertTrue(
          adjacencyGraph.hasNode(secondGridCoord),
          String.format("Node %s is missing", secondGridCoord));
      assertTrue(
          adjacencyGraph
              .getNeighbors(firstGridCoord)
              .contains(new Edge<GridCoord>(secondGridCoord, 1)),
          String.format("coord %s should be linked to %s", firstGridCoord, secondGridCoord));
      assertTrue(
          adjacencyGraph
              .getNeighbors(secondGridCoord)
              .contains(new Edge<GridCoord>(firstGridCoord, 1)),
          String.format("coord %s should be linked back to %s", secondGridCoord, firstGridCoord));
    }
  }

  private static FieldSquare[][] placeObstructions() {
    FieldSquare[][] fieldSquares = new FieldSquare[Pathing.CELL_X_MAX][Pathing.CELL_Y_MAX];

    for (int x = 0; x < Pathing.CELL_X_MAX; x++) {
      for (int y = 0; y < Pathing.CELL_Y_MAX; y++) {
        final double xCoord = x * Pathing.CELL_SIZE_METERS;
        final double yCoord = y * Pathing.CELL_SIZE_METERS;
        fieldSquares[x][y] =
            FieldObstructionMap.isInsideObstruction(new Translation2d(xCoord, yCoord))
                ? FieldSquare.OBSTRUCTION
                : FieldSquare.EMPTY;
      }
    }

    return fieldSquares;
  }

  private Expect expect;

  @UtilTest
  public void internalCollisionGridMatchesSnapshot() {
    // use reflection to read the internal BoolGrid collisionGrid
    Optional<BoolGrid> optCollisionGrid = Optional.empty();
    Optional<BoolGrid> optDangerGrid = Optional.empty();
    try {
      Field field1 = RubenManueverGenerator.class.getDeclaredField("collisionGrid");
      field1.setAccessible(true);
      Field field2 = RubenManueverGenerator.class.getDeclaredField("dangerGrid");
      field2.setAccessible(true);

      var obj1 = field1.get(new RubenManueverGenerator());
      var obj2 = field2.get(new RubenManueverGenerator());

      optCollisionGrid = obj1 instanceof BoolGrid ? Optional.of((BoolGrid) obj1) : Optional.empty();
      optDangerGrid = obj2 instanceof BoolGrid ? Optional.of((BoolGrid) obj2) : Optional.empty();
    } catch (NoSuchFieldException | IllegalAccessException e) {
      e.printStackTrace();
    }

    if (optCollisionGrid.isEmpty() || optDangerGrid.isEmpty()) {
      fail("grids are empty because reflection failed");
    }

    FieldSquare[][] fieldSquares = new FieldSquare[Pathing.CELL_X_MAX][Pathing.CELL_Y_MAX];

    var collisionGrid = optCollisionGrid.get();
    var dangerGrid = optDangerGrid.get();

    for (int x = 0; x < Pathing.CELL_X_MAX; x++) {
      for (int y = 0; y < Pathing.CELL_Y_MAX; y++) {
        // if the collision grid is true, the field square should be an obstruction
        if (collisionGrid.get(x, y)) {
          fieldSquares[x][y] = FieldSquare.OBSTRUCTION;
        } else if (dangerGrid.get(x, y)) {
          fieldSquares[x][y] = FieldSquare.DANGER;
        } else {
          fieldSquares[x][y] = FieldSquare.EMPTY;
        }
      }
    }

    var displayCoords =
        List.of(
            new GridCoord(10, 62),
            new GridCoord(32, 10),
            new GridCoord(new Translation2d(1.8, .5)),
            new GridCoord(new Translation2d(14.75, .5)));

    for (var coord : displayCoords) {
      fieldSquares[coord.x][coord.y] = FieldSquare.START;
    }

    StringBuilder sb = new StringBuilder();
    DisplayFieldArray.renderField(sb, fieldSquares);

    ObjectMapper mapper = new ObjectMapper();
    mapper.enable(SerializationFeature.INDENT_OUTPUT);

    for (var coord : displayCoords) {
      try {
        sb.append(mapper.writeValueAsString(coord));
      } catch (JsonProcessingException e) {
        sb.append("failed to serialize coord\n");
      }
    }

    expect.toMatchSnapshot(sb.toString());
  }

  public static Stream<Arguments> findFullPathMatchesSnapshotProvider() {
    return Stream.of(
        Arguments.of(new GridCoord(50, 50), new GridCoord(50, 55)),
        Arguments.of(new GridCoord(10, 62), new GridCoord(32, 10))
        // load bearing comment (hold the final brace)
        );
  }

  @UtilParamTest
  @MethodSource("findFullPathMatchesSnapshotProvider")
  public void findFullPathMatchesSnapshot(GridCoord start, GridCoord end) {
    RubenManueverGenerator rubenManueverGenerator = new RubenManueverGenerator();

    FieldSquare[][] fieldSquares = placeObstructions();

    var path = rubenManueverGenerator.findFullPath(start, end);

    for (var coord : path.get()) {
      fieldSquares[coord.x][coord.y] = FieldSquare.PATH;
    }

    fieldSquares[start.x][start.y] = FieldSquare.START;
    fieldSquares[end.x][end.y] = FieldSquare.END;

    StringBuilder sb = new StringBuilder();
    DisplayFieldArray.renderField(sb, fieldSquares);

    expect
        .scenario(String.format("%s -> %s", start.toString(), end.toString()))
        .toMatchSnapshot(sb.toString());
  }

  public static Stream<Arguments> findPathAndCriticalPointsMatchesSnapshotProvider() {
    return Stream.of(
        Arguments.of(new GridCoord(50, 50), new GridCoord(50, 55)),
        Arguments.of(new GridCoord(10, 62), new GridCoord(32, 10))
        // load bearing comment (hold the final brace)
        );
  }

  @UtilParamTest
  @MethodSource("findPathAndCriticalPointsMatchesSnapshotProvider")
  public void findPathAndCriticalPointsMatchesSnapshot(GridCoord start, GridCoord end) {
    RubenManueverGenerator rubenManueverGenerator = new RubenManueverGenerator();

    FieldSquare[][] fieldSquares = placeObstructions();

    var path = rubenManueverGenerator.findFullPath(start, end);
    var criticalPoints = RubenManueverGenerator.findCriticalPoints(path.get());

    for (var coord : path.get()) {
      fieldSquares[coord.x][coord.y] = FieldSquare.PATH;
    }

    for (var coord : criticalPoints) {
      fieldSquares[coord.x][coord.y] = FieldSquare.CRITICAL_POINT;
    }

    // fieldSquares[start.x][start.y] = FieldSquare.START;
    // fieldSquares[end.x][end.y] = FieldSquare.END;

    StringBuilder sb = new StringBuilder();
    DisplayFieldArray.renderField(sb, fieldSquares);

    expect
        .scenario(
            String.format(
                "%s -> %s len: %s",
                start.toString(), end.toString(), path.map(List::size).orElse(-1)))
        .toMatchSnapshot(sb.toString());
  }

  public static Stream<Arguments> findRedundantCriticalPointsProvider() {
    return Stream.of(
        Arguments.of(new GridCoord(50, 50), new GridCoord(50, 55)),
        Arguments.of(new GridCoord(10, 62), new GridCoord(32, 10)),
        Arguments.of(
            new GridCoord(new Translation2d(14.94, 6.78)),
            new GridCoord(new Translation2d(2.34, .78)))
        // load bearing comment (hold the final brace)
        );
  }

  @UtilParamTest
  @MethodSource("findRedundantCriticalPointsProvider")
  public void findRedundantCriticalPoints(GridCoord start, GridCoord end) {
    RubenManueverGenerator rubenManueverGenerator = new RubenManueverGenerator();

    FieldSquare[][] fieldSquares = placeObstructions();

    var path = rubenManueverGenerator.findFullPath(start, end);
    var criticalPoints = RubenManueverGenerator.findCriticalPoints(path.get());
    var neededCriticalPoints = rubenManueverGenerator.simplifyCriticalPoints(criticalPoints);

    for (var coord : path.get()) {
      fieldSquares[coord.x][coord.y] = FieldSquare.PATH;
    }

    var removedCriticalPoints = new HashSet<>(criticalPoints);
    removedCriticalPoints.removeAll(neededCriticalPoints);

    for (var coord : removedCriticalPoints) {
      fieldSquares[coord.x][coord.y] = FieldSquare.REDUNDANT_CRITICAL_POINT;
    }

    for (var coord : neededCriticalPoints) {
      fieldSquares[coord.x][coord.y] = FieldSquare.CRITICAL_POINT;
    }

    StringBuilder sb = new StringBuilder();
    DisplayFieldArray.renderField(sb, fieldSquares);

    expect
        .scenario(
            String.format(
                "%s -> %s len: %s",
                start.toString(), end.toString(), path.map(List::size).orElse(-1)))
        .toMatchSnapshot(sb.toString());
  }

  public static Stream<Arguments> computePathPointsForSplineProvider() {
    return Stream.of(
        Arguments.of(new GridCoord(50, 50), new GridCoord(50, 55)),
        Arguments.of(new GridCoord(10, 62), new GridCoord(32, 10)),
        Arguments.of(
            new GridCoord(new Translation2d(14.94, 6.78)),
            new GridCoord(new Translation2d(2.34, .78))),
        Arguments.of(new GridCoord(71, 46), new GridCoord(new Translation2d(1.8, .5))),
        Arguments.of(
            new GridCoord(new Translation2d(1.8, .5)), new GridCoord(new Translation2d(1.8, 4.6))),
        Arguments.of(new GridCoord(78, 75), new GridCoord(18, 16))
        // load bearing comment (hold the final brace)
        );
  }

  @UtilParamTest
  @MethodSource("computePathPointsForSplineProvider")
  public void computePathPointsForSplineMatchesSnapshot(GridCoord start, GridCoord end) {
    RubenManueverGenerator rubenManueverGenerator = new RubenManueverGenerator();

    FieldSquare[][] fieldSquares = placeObstructions();

    var path = rubenManueverGenerator.findFullPath(start, end);
    if (path.isEmpty()) {
      fail(String.format("No path found from %s to %s", start, end));
    }
    var criticalPoints = RubenManueverGenerator.findCriticalPoints(path.get());
    var neededCriticalPoints = rubenManueverGenerator.simplifyCriticalPoints(criticalPoints);
    var pathPoints =
        RubenManueverGenerator.computePathPointsFromCriticalPoints(
            neededCriticalPoints,
            new Pose2d(start.toTranslation2d(), new Rotation2d()),
            new ChassisSpeeds(),
            new Pose2d(end.toTranslation2d(), new Rotation2d()));

    for (var coord : path.get()) {
      fieldSquares[coord.x][coord.y] = FieldSquare.PATH;
    }

    // build a spline
    PathPlannerTrajectory trajectory =
        PathPlanner.generatePath(new PathConstraints(3, 1), pathPoints);

    // draw the spline
    for (var states : trajectory.getStates()) {
      var coord = new GridCoord(states.poseMeters.getTranslation());
      fieldSquares[coord.x][coord.y] = FieldSquare.SPLINE;
    }

    for (var coord : neededCriticalPoints) {
      fieldSquares[coord.x][coord.y] = FieldSquare.CRITICAL_POINT;
    }

    // fieldSquares[start.x][start.y] = FieldSquare.START;
    // fieldSquares[end.x][end.y] = FieldSquare.END;

    StringBuilder sb = new StringBuilder();
    DisplayFieldArray.renderField(sb, fieldSquares);

    // write the pathpoints to the sb
    for (var point : pathPoints) {
      // use reflection to set the access and get the position, heading, and holonomic rotation
      try {
        Field positionField = point.getClass().getDeclaredField("position");
        positionField.setAccessible(true);
        Translation2d position = (Translation2d) positionField.get(point);

        Field headingField = point.getClass().getDeclaredField("heading");
        headingField.setAccessible(true);
        Rotation2d heading = (Rotation2d) headingField.get(point);

        Field holonomicRotationField = point.getClass().getDeclaredField("holonomicRotation");
        holonomicRotationField.setAccessible(true);
        Rotation2d holonomicRotation = (Rotation2d) holonomicRotationField.get(point);

        sb.append(
            String.format("pos:%s heading:%s rotation:%s\n", position, heading, holonomicRotation));
      } catch (NoSuchFieldException | IllegalAccessException e) {
        sb.append("error getting field\n");
      }
    }

    expect
        .scenario(
            String.format(
                "%s -> %s len: %s",
                start.toString(), end.toString(), path.map(List::size).orElse(-1)))
        .toMatchSnapshot(sb.toString());
  }

  @UtilTest
  public void stepsShouldNotIntroduceMoreThanTwoPointsIntoStraightLinePath() {
    var start = new GridCoord(50, 50);
    var end = new GridCoord(50, 55);

    RubenManueverGenerator rubenManueverGenerator = new RubenManueverGenerator();

    var fullPath = rubenManueverGenerator.findFullPath(start, end);
    assertIterableEquals(
        List.of(
            new GridCoord(50, 50),
            new GridCoord(50, 51),
            new GridCoord(50, 52),
            new GridCoord(50, 53),
            new GridCoord(50, 54),
            new GridCoord(50, 55)),
        fullPath.get());

    var criticalPoints = RubenManueverGenerator.findCriticalPoints(fullPath.get());
    assertIterableEquals(List.of(new GridCoord(50, 50), new GridCoord(50, 55)), criticalPoints);

    var simplifiedCriticalPoints = rubenManueverGenerator.simplifyCriticalPoints(criticalPoints);
    assertIterableEquals(List.of(new GridCoord(50, 50), new GridCoord(50, 55)), criticalPoints);

    var pathPoints =
        RubenManueverGenerator.computePathPointsFromCriticalPoints(
            simplifiedCriticalPoints,
            new Pose2d(start.toTranslation2d(), new Rotation2d()),
            new ChassisSpeeds(),
            new Pose2d(end.toTranslation2d(), new Rotation2d()));
    assertEquals(2, pathPoints.size(), "should only have start and end");
  }

  @UtilTest
  public void computePathMatchesSnapshot() {
    var start = new Pose2d(new GridCoord(78, 75).toTranslation2d(), new Rotation2d());
    var chassisSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(-1, -1, 0, new Rotation2d());
    var end = new Pose2d(new GridCoord(18, 16).toTranslation2d(), new Rotation2d());

    var rubenManueverGenerator = new RubenManueverGenerator();

    var path =
        rubenManueverGenerator
            .computePath(start, chassisSpeeds, end, new PathConstraints(3, 1))
            .get();

    var fieldSquares = placeObstructions();

    for (var states : path.getStates()) {
      var coord = new GridCoord(states.poseMeters.getTranslation());
      fieldSquares[coord.x][coord.y] = FieldSquare.SPLINE;
    }

    var startGridCoord = new GridCoord(start.getTranslation());
    var endGridCoord = new GridCoord(end.getTranslation());
    fieldSquares[startGridCoord.x][startGridCoord.y] = FieldSquare.START;
    fieldSquares[endGridCoord.x][endGridCoord.y] = FieldSquare.END;

    StringBuilder sb = new StringBuilder();
    DisplayFieldArray.renderField(sb, fieldSquares);

    expect.toMatchSnapshot(sb.toString());
  }
}
