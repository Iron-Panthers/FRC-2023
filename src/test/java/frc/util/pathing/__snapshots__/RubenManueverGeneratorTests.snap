frc.util.pathing.RubenManueverGeneratorTests.findFullPathMatchesSnapshot[Translation2d(X: 15.40, Y: 7.40) -> Translation2d(X: 15.40, Y: 7.50)]=[
Optional[[Translation2d(X: 15.40, Y: 7.40), Translation2d(X: 15.40, Y: 7.50)]]
]